resource reference resolver

geoberle opened the pull request:
### Problem

resource files are somehow detached from our regular graphql types. they are just references as a path and it is up to the respective integration to fetch the content of such a resource with a dedicated query. this consumes a lot of time during integration runs and pr checks, because such resources are fetched one by one (see terraform-resources defaults or jjb config_path).

### Implementation idea

`isResource`field definition enhanced with `resolveResource: true` will be queryable as graphql type `Resource_v1`

### Example

schema.yml

```yaml
- name: JenkinsConfig_v1
  fields:
    ...
    - { name: config_path, type: string, isResource: true, resolveResource: true }
```

query

```
{
  jenkins_configs: jenkins_configs_v1 {
    name
    ...
    type
    config
    config_path {
      content
      sha256sum
    }
  }
}
```

ref: https://issues.redhat.com/browse/APPSRE-5950

Signed-off-by: Gerd Oberlechner <goberlec@redhat.com>

steveteahan commented:
Do we have any tests in this repo for the new functionality?

I've done little work in qontract-server, so not the best person to review this. With that being said, strong +1 to the idea because it reduces the number of GraphQL calls and simplifies the client code a bit.

geoberle commented:
> Do we have any tests in this repo for the new functionality?
> 
> I've done little work in qontract-server, so not the best person to review this. With that being said, strong +1 to the idea because it reduces the number of GraphQL calls and simplifies the client code a bit.

thank you. i wanted your feedback from a pure consumer perspective to know if i should go forward with this.

i will ask for reviews once i have tests

bhushanthakur93 commented:
Just note that we will need to update json-schema for GraphQL to reflect this new field `resolveResource` .

geoberle commented:
> Just note that we will need to update json-schema for GraphQL to reflect this new field `resolveResource` .
how could i forget :) 

https://github.com/app-sre/qontract-schemas/pull/186

bhushanthakur93 commented:
> added tests in [54381e1](https://github.com/app-sre/qontract-server/pull/147/commits/54381e1e1462217a1ae67f6bb7a3ff24367bbcd8)
> 
> the tests use the new bundle format with `confs` and i'm not going to add tests for the old format anymore @bhushanthakur93 is that ok?

Yes, that sounds good.

geoberle commented:
> @geoberle just curious, what were the downsides of using `isResource` flag in this case? Could we have possibly change the existing type or add a new type instead?

i considered it for a moment but a current use of `isResource` is more like a marker for a graphql type than a marker for a field.

the goal would be to make `resolveResource=true` the default in the future. for now it helps us to keep non-resolving behaviour so we can change integration by integration to use this capability. and if some integration is really just interested in the path of a resource there is always the possibility to write a query like this

```
{
  jenkins_configs: jenkins_configs_v1 {
    name
    ...
    type
    config
    config_path {
      path <----
    }
  }
}
```

does that make sense?

bhushanthakur93 commented:
> > @geoberle just curious, what were the downsides of using `isResource` flag in this case? Could we have possibly change the existing type or add a new type instead?
> 
> i considered it for a moment but a current use of `isResource` is more like a marker for a graphql type than a marker for a field.
Not sure I'm following this entirely. `isResource` looks like a marker defined on the field and fields are essentially GraphQL types. Is your point coming from the fact that only `Resource_v1` field has `isResource` in query? 
> the goal would be to make `resolveResource=true` the default in the future. for now it helps us to keep non-resolving behaviour so we can change integration by integration to use this capability. and if some integration is really just interested in the path of a resource there is always the possibility to write a query like this
> 
> ```
> {
>   jenkins_configs: jenkins_configs_v1 {
>     name
>     ...
>     type
>     config
>     config_path {
>       path <----
>     }
>   }
> }
> ```
> 
> does that make sense?

The use-case makes sense to me. I'm a little bit confused of similar sounding attributes `isResource`, `isResourceRef` etc and their usage but I think we can address that concern by adding the details in schema documentation MR in qontract-server. You are more than welcome to contribute there :smile:

geoberle commented:
isResourceRef was introduced as marker to know that a field is a string pointer to a resource file without changing the schema. A field with isResourceRef is still a String and does not require any change in app-interface or qontract-reconcile. This information is used during bundling to check for stale refs and creates backrefs. 

But I have an idea how to revamp that a bit so we can use isResource in both contexts. Let's discuss tomorrow.

geoberle commented:
@bhushanthakur93 i updated the proposal to only use `isResource` and eliminated `isResourceRef` because the distinction between a real resource and just a ref can be made via the type of the field (`Resource_v1` vs `string`).

i updated the example in the PR description.

see also https://github.com/app-sre/qontract-schemas/pull/186 (removes isResourceRef and adds resolveResource)
and https://github.com/app-sre/qontract-validator/pull/40 (updated a fixture file there)
